sql/statistics.go

```diff
@@ -26,19 +30,127 @@ type StatisticsTable interface {
 }
+
+// Statistic is the top-level interface for accessing cardinality and
+// costing estimates for an index prefix.
+type Statistic interface {
+	JSONWrapper
+	RowCount() uint64
+	DistinctCount() uint64
+	NullCount() uint64
+	AvgSize() uint64
+	CreatedAt() time.Time
+	Columns() []string
+	Types() []Type
+	Qualifier() StatQualifier
+	Histogram() Histogram
+}
+
+func NewQualifierFromString(q string) (StatQualifier, error) {
+	parts := strings.Split(q, ".")
+	if len(parts) < 3 {
+		return StatQualifier{}, fmt.Errorf("invalid qualifier string: '%s', expected '<database>.<table>.<index>'", q)
+	}
+	return StatQualifier{db: parts[0], table: parts[1], index: parts[2]}, nil
+}
+
+func NewStatQualifier(db, table, index string) StatQualifier {
+	return StatQualifier{db: db, table: table, index: index}
+}
+
+// StatQualifier is the namespace hierarchy for a given statistic.
+// The qualifier and set of columns completely describes a unique stat.
+type StatQualifier struct {
+	db    string
+	table string
+	index string
+}
+
+func (q StatQualifier) String() string {
+	if q.index != "" {
+		return fmt.Sprintf("%s.%s.%s", q.db, q.table, q.index)
+	}
+	return fmt.Sprintf("%s.%s", q.db, q.table)
+}
+
+func (q StatQualifier) Db() string {
+	return q.db
+}
+
+func (q StatQualifier) Table() string {
+	return q.table
+}
+
+func (q StatQualifier) Index() string {
+	return q.index
+}
+
+// Histogram is a collection of non-overlapping buckets that
+// estimate the costing statistics for an index prefix.
+// Note that a non-unique key can cross bucket boundaries.
+type Histogram []HistogramBucket
+
+func (h Histogram) ToInterface() interface{} {
+	ret := make([]interface{}, len(h))
+	for i, b := range h {
+		upperBound := make([]interface{}, len(b.UpperBound()))
+		for i, v := range b.UpperBound() {
+			upperBound[i] = v
+		}
+		mcvs := make([][]interface{}, len(b.Mcvs()))
+		for i, mcv := range b.Mcvs() {
+			mcvs[i] = make([]interface{}, len(mcv))
+			for j, v := range mcv {
+				mcvs[i][j] = v
+			}
+		}
+		ret[i] = map[string]interface{}{
+			"row_count":      b.RowCount(),
+			"null_count":     b.NullCount(),
+			"distinct_count": b.DistinctCount(),
+			"bound_count":    b.BoundCount(),
+			"mcv_counts":     b.McvCounts(),
+			"mcvs":           mcvs,
+			"upper_bound":    upperBound,
+		}
+	}
+	return ret
+}
+
+// HistogramBucket contains statistics for a fragment of an
+// index's keyspace.
+type HistogramBucket interface {
+	RowCount() uint64
+	DistinctCount() uint64
+	NullCount() uint64
+	BoundCount() uint64
+	UpperBound() Row
+	McvCounts() []uint64
+	Mcvs() []Row
+}
+
+// JSONWrapper is an integrator specific implementation of a JSON field value.
+// The query engine can utilize these optimized access methods improve performance
+// by minimizing the need to unmarshall a JSONWrapper into a JSONDocument.
+type JSONWrapper interface {
+	// ToInterface converts a JSONWrapper to an interface{} of simple types
+	ToInterface() interface{}
+}
```

Comment:
(minor) This interface seems like a better fit closer to the other JSON types. Does putting it in `json.go` in the `sql/types` package cause import cycles?

Reply:
yeah this was an import cycle, i want statistics to be json but `sql` can't depend on `types`